src/harness/sas_objects.py

```diff
@@ -0,0 +1,387 @@
+                if heartbeat_response['response']['responseCode'] == 0:
+                    transmit_expire_time = datetime.strptime(heartbeat_response['transmitExpireTime'],
+                                                             '%Y-%m-%dT%H:%M:%SZ')
+                    self.testcase.assertLess(datetime.utcnow(), transmit_expire_time)
+                    self.testcase.assertLessEqual(
+                        (transmit_expire_time - datetime.utcnow()).total_seconds(), 240)
+                    grant_object.setState(Grant.AUTHORIZED)
+                else:
+                    self.mapResponseCodeToGrantState(
+                        heartbeat_response['response']['responseCode'], grant_object)
+
+                    #If any heartbeats contain suggested operating params,
+                    #relinquish the corresponding grant
+                    if 'operationParam' in heartbeat_response:
+                        relinquish_request = {
+                            'relinquishmentRequest': [{
+                                'cbsdId': grant_object.getGrantRequest()['cbsdId'],
+                                'grantId': grant_object.getGrantRequest()['grantId']
+                            }]
+                        }
+                        self.testcase._sas.Relinquishment(
+                            relinquish_request, self.ssl_cert, self.ssl_key)
+                        heartbeat_responses_with_operation_parameters.update(
+                            {heartbeat_response['cbsdId']: heartbeat_response})
+                        del cbsd_object_item.grant_objects[heartbeat_response['grant_id']]
+
+            #For all grants in the previous step, request a new grant with the
+            #suggested operation parameters
+            for cbsd_id, heartbeatResponse in heartbeat_responses_with_operation_parameters.items():
+                grant_0 = json.load(
+                    open(os.path.join('testcases', 'testdata', 'grant_0.json')))
+                grant_0['cbsdId'] = cbsd_id
+                grant_0['operationParam']['maxEirp'] = heartbeatResponse['operationParam']['maxEirp']
+                grant_0['operationParam']['operationFrequencyRange']['lowFrequency'] = \
+                    heartbeatResponse['operationParam']['operationFrequencyRange']['lowFrequency']
+                grant_0['operationParam']['operationFrequencyRange']['highFrequency'] = \
+                    heartbeatResponse['operationParam']['operationFrequencyRange']['highFrequency']
+                request = {'grantRequest': [grant_0]}
+                response = self.testcase._sas.Grant(request , self.ssl_cert, self.ssl_key)[
+                    'grantResponse'][0]
+                self.cbsd_objects.grant_objects.update(
+                    {response['grantId']: request })
+                if response['response']['responseCode'] == "0":
+                    #Heartbeat all active grants
+                    self.heartbeatRequestForAllGrants()
```

Comment:
Why is it this inside the `for` loop? Please check the indentation level of this and the previous code.

Reply:
Logic   has been changed